openedx_tagging/core/tagging/rest_api/v1/views.py

```diff
@@ -654,34 +659,40 @@ def get_queryset(self) -> TagDataQuerySet:
         parent_tag_value = self.request.query_params.get("parent_tag", None)
-        root_only = "root_only" in self.request.query_params
         include_counts = "include_counts" in self.request.query_params
         search_term = self.request.query_params.get("search_term", None)
-
-        if parent_tag_value:
-            # Fetching tags below a certain parent is always paginated and only returns the direct children
-            depth = 1
-            if root_only:
-                raise ValidationError("?root_only and ?parent_tag cannot be used together")
+        full_depth_threshold = int(self.request.query_params.get("full_depth_threshold", 0))
+        if full_depth_threshold < 0 or full_depth_threshold > MAX_FULL_DEPTH_THRESHOLD:
+            raise ValidationError("Invalid full_depth_threshold")
+
+        if full_depth_threshold or search_term:
+            # If full_depth_threshold is set, default to maximum depth and later prune to depth=1 if needed.
+            # We also need to do a deep search if there is a search term, to ensure we return parent items with children
+            # that match.
+            depth = None
         else:
-            if root_only:
-                depth = 1  # User Explicitly requested to load only the root tags for now
-            elif search_term:
-                depth = None  # For search, default to maximum depth but use normal pagination
-            elif taxonomy.tag_set.count() > TAGS_THRESHOLD:
-                # This is a very large taxonomy. Only load the root tags at first, so users can choose what to load.
-                depth = 1
-            else:
-                # We can load and display all the tags in the taxonomy at once:
-                self.pagination_class = DisabledTagsPagination
-                depth = None  # Maximum depth
+            depth = 1  # Otherwise (default), load a single level of results only.
 
-        return taxonomy.get_filtered_tags(
+        results = taxonomy.get_filtered_tags(
             parent_tag_value=parent_tag_value,
             search_term=search_term,
             depth=depth,
             include_counts=include_counts,
         )
+        if depth == 1:
+            # We're already returning just a single level. It will be paginated normally.
+            return results
+        elif full_depth_threshold and len(results) < full_depth_threshold:
+            # We can load and display all the tags in this (sub)tree at once:
+            self.pagination_class = DisabledTagsPagination
+            return results
+        else:
+            # We had to do a deep query, but we will only return one level of results.
+            # This is because the user did not request a deep response (via full_depth_threshold) or the result was too
+            # large (larger than the threshold).
+            # It will be paginated normally.
+            min_depth = 0 if parent_tag_value is None else results[0]["depth"]
```

Comment:
@bradenmacdonald I tried out the updated code, is now returns the results correctly. However when I try to fetch subtags where there is no more results expected, I get an index out of range here (because there are no more relevant search matches) We need to handle that case.

To reproduce: 

`http://localhost:18010/api/content_tagging/v1/taxonomies/14/tags/?parent_tag=Sensory+Abilities&search_term=sensory`

Returns the last tag that contains the search result, however it also includes a sub_tags_url:

```json
{
  "next":null,
  "previous":null,
  "count":1,
  "num_pages":1,
  "current_page":1,
  "start":0,
  "results":[
    {
      "value":"Other Sensory Abilities",
      "external_id":"A04c",
      "child_count":4,
      "depth":2,
      "parent_value":"Sensory Abilities",
      "_id":626321,
      "sub_tags_url":"http://localhost:18010/api/content_tagging/v1/taxonomies/14/tags/?parent_tag=Other+Sensory+Abilities&search_term=sensory"
    }
  ]
}
```
If I try to access the sub_tag_url provided I face the index out of  range issue:

`http://localhost:18010/api/content_tagging/v1/taxonomies/14/tags/?parent_tag=Other+Sensory+Abilities&search_term=sensory`

There are no more sub tags that match "sensory" under "Other Sensory Abilities". I think we can either not return a sub_tags_url if there are no more further results, or if its easier, just add a check before accessing the index and returning an empty result.

Reply:
@yusuf-musleh Thanks, I'll see if I can figure something out for that. But just to be clear, your initial editing PR won't use the `search_term` feature right? So it wouldn't be affected by this bug.

Reply:
@yusuf-musleh I fixed that bug. Now it won't return an error if you try to get the children of a leaf node, and it also won't provide a `sub_tags_url` for those tags in the first place. Also, the `child_count` is now accurate for the search results.

Reply:
@bradenmacdonald 

> But just to be clear, your initial editing PR won't use the search_term feature right? So it wouldn't be affected by this bug.

Yup, that's right.

> I fixed that bug. Now it won't return an error if you try to get the children of a leaf node, and it also won't provide a sub_tags_url for those tags in the first place. Also, the child_count is now accurate for the search results.

It working great now as expected, thanks!